equinox/nn/attention.py

```diff
@@ -11,6 +11,183 @@
+            attn_weights, key=key, inference=inference, deterministic=deterministic
+        )
+
+    return attn_weights
+
+
+def dot_product_attention(
+    query: Array,
+    key_: Array,
+    value: Array,
+    bias: Optional[Array] = None,
+    mask: Optional[Array] = None,
+    dropout_p: float = 0.0,
+    *,
+    dropout: Optional[Dropout] = None,
+    key: Optional["jax.random.PRNGKey"] = None,
+    inference: Optional[bool] = None,
+    deterministic: Optional[bool] = None,
+):
+
+    r"""Computes dot-product attention given query, key, and value.
+    This is the core function for applying attention based on
+    https://arxiv.org/abs/1706.03762.
+
+    - $\text{Attention}$ is defined as
+      $\text{Attention}(\widetilde{Q}, \widetilde{K}, \widetilde{V})
+       = \text{softmax}(\frac{\widetilde{Q}\widetilde{K}^\intercal}
+                             {\sqrt{d_\text{qk}}})\widetilde{V}$.
+
+    This also supports multi-query attention (https://arxiv.org/pdf/1911.02150).
+    **Arguments:**
+
+    - `query`: Query vectors. Should be a JAX array of shape
+            `(batch ... query_seq_length, num_heads, qk_size)`.
+    - `key_`: Key vectors. Should be a JAX array of shape
+            `(batch ... kv_seq_length, num_heads, qk_size)`.
+    - `value`: Value vectors. Should be a JAX array of shape
+            `(batch ... kv_seq_length, num_heads, vo_size)`.
+    - `mask`: Optional mask preventing attention to certain positions. Should be a
+            JAX array of shape `(num_heads, query_seq_length, kv_seq_length)`.
+    - `dropout_p`: Dropout probability on attention weights.
+    - `dropout`: Already initialized Dropout module. Unused if `dropout_p != 0.`
+          (Keyword only argument).
+    - `key`: A `jax.random.PRNGKey` used for dropout. Unused if `dropout_p = None`.
+            (Keyword only argument.)
+    - `inference`: As [`equinox.nn.Dropout.__call__`][]. (Keyword only
+            argument.)
+    - `deterministic`: (Deprecated in favour of `inference`.)
+
+    **Returns:**
+
+    A JAX array of shape `(batch..., query_seq_length, num_heads, vo_size)`.
+    """
+    assert key_.ndim == value.ndim, "k, v must have same rank."
+    if query.ndim == key_.ndim == value.ndim:
+        assert (
+            query.shape[:-3] == key_.shape[:-3] == value.shape[:-3]
+        ), "q, k, v batch dims must match."
+        assert (
+            query.shape[-2] == key_.shape[-2] == value.shape[-2]
+        ), "q, k, v num_heads must match."
+        assert key_.shape[-3] == value.shape[-3], "k, v lengths must match"
+    elif query.ndim > key_.ndim and query.ndim > value.ndim:
+        # support for multi-query attention
+        assert (
+            query.shape[:-3] == key_.shape[:-2] == value.shape[:-2]
+        ), "q, k batch dims must match."
+        assert key_.shape[-2] == value.shape[-2], "k, v lengths must match"
+    else:
+        raise ValueError("q must have equal or more dimensions than k, v.")
```

Comment:
Probably it won't matter after the simplification from the above comments anyway, but I think these checks are pretty much duplication the ones in `dot_product_attention_weights`, so things can be simplified.

Reply:
Yeah I simplified to only check that key and value shapes match (since query and key are checked in the weights function).